haskell/private/actions/compile.bzl

```diff
@@ -68,21 +68,20 @@ def _process_hsc_file(hs, cc, ghc_defs_dump, hsc_file):
 
   Args:
     hs: Haskell context.
+    cc: CcInteropInfo, information about C dependencies.
     ghc_defs_dump: File with GHC definitions.
     hsc_file: hsc file to process.
 
   Returns:
     File: Haskell source file created by processing hsc_file.
   """
 
-  hsc_output_dir = hs.actions.declare_directory(
```

Comment:
Creating a per-module unique directory to put a file (which already has unique-enough path) is awkward. I first made this change when I just started and did not better. The proper solution is avoid declaring directory and instead add raw string prefix to each file before declaring them. This way they all have the prefix, but Bazel can't nitpick on the basis that some output path is prefix of another (not true!). This requires a minor change to the `declare_compiled` helper to make it accept strings instead of `File`s.